src/gnc_environment.cpp

```diff
@@ -0,0 +1,102 @@
+//
+// src/gnc_environment.hpp
+// PSim
+//
+// Contributors:
+//   Kyle Krol  kpk63@cornell.edu
+//
+// Pathfinder for Autonomous Navigation
+// Space Systems Design Studio
+// Cornell Univeristy
+//
+
+#include <gnc_constants.hpp>
+#include <gnc_environment.hpp>
+#include <gnc_utilities.hpp>
+
+#include <cmath>
+
+#include "inl/geomag.inl"
+
+namespace gnc {
+namespace env {
+
+void earth_attitude(double t, lin::Vector4d &q) {
+  using namespace lin;
+  // Determine a transformation from ecef0 to ecef0p
+  Vector4d q_ecef0p_ecef0({
+    t * constant::earth_precession_rate(0),
+    t * constant::earth_precession_rate(1),
+    t * constant::earth_precession_rate(2),
+    1.0
+  });
+  q_ecef0p_ecef0 = q_ecef0p_ecef0 / norm(q_ecef0p_ecef0);
+  // Earth's rotation angle about the z axis (assumes x, y components are zero)
+  double theta = t * constant::earth_rate_ecef_z;
+  Vector4d q_ecef_ecef0p({
+    0.0,
+    0.0,
+    sin(theta / 2.0),
+    cos(theta / 2.0)
+  });
+  // Rotate through ECI -> ECEF0 -> ECEF0P -> ECEF
+  Vector4d temp;
+  utl::quat_cross_mult(q_ecef0p_ecef0, constant::q_ecef0_eci, temp);
+  utl::quat_cross_mult(q_ecef_ecef0p, temp, q);
+}
+
+void earth_attitude(double t, lin::Vector4f &q) {
+  lin::Vector4d _q;
+  earth_attitude(t, _q);
+  for (lin::size_t i = 0; i < q.size(); i++)
+      q(i) = static_cast<float>(_q(i));
+}
+
+void earth_angular_rate(double t, lin::Vector3d &w) {
+  w = { 0.0, 0.0, constant::earth_rate_ecef_z };
+}
```

Comment:
@nhz2 I was curious about this bit of code from MATLAB. Doesn't this neglect the `w` contribution from the Earth's precession? Are we just calling it negligible because it's so small relative to the rotation about the z axis?